lib/icingadb/icingadbchecktask.cpp

```diff
@@ -0,0 +1,449 @@
+{
+	CheckCommand::Ptr commandObj = CheckCommand::ExecuteOverride ? CheckCommand::ExecuteOverride : checkable->GetCheckCommand();
+
+	Host::Ptr host;
+	Service::Ptr service;
+	tie(host, service) = GetHostService(checkable);
+
+	MacroProcessor::ResolverList resolvers;
+	String silenceMissingMacroWarning;
+
+	if (MacroResolver::OverrideMacros)
+		resolvers.emplace_back("override", MacroResolver::OverrideMacros);
+
+	if (service)
+		resolvers.emplace_back("service", service);
+	resolvers.emplace_back("host", host);
+	resolvers.emplace_back("command", commandObj);
+	resolvers.emplace_back("icinga", IcingaApplication::GetInstance());
+
+	auto resolve ([&](const String& macro) {
+		return MacroProcessor::ResolveMacros(macro, resolvers, checkable->GetLastCheckResult(),
+			&silenceMissingMacroWarning, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros);
+	});
+
+	struct Thresholds
+	{
+		Value Warning, Critical;
+	};
+
+	auto resolveThresholds ([&resolve](const String& wmacro, const String& cmacro) {
+		return Thresholds{resolve(wmacro), resolve(cmacro)};
+	});
+
+	String icingadbName = resolve("$icingadb_name$");
+
+	auto dumpTakesThresholds (resolveThresholds("$icingadb_dumptakes_warning$", "$icingadb_dumptakes_critical$"));
+	auto syncTakesThresholds (resolveThresholds("$icingadb_synctakes_warning$", "$icingadb_synctakes_critical$"));
+	auto icingaBacklogThresholds (resolveThresholds("$icingadb_redis_backlog_warning$", "$icingadb_redis_backlog_critical$"));
+	auto icingadbBacklogThresholds (resolveThresholds("$icingadb_rdbms_backlog_warning$", "$icingadb_rdbms_backlog_critical$"));
+
+	if (resolvedMacros && !useResolvedMacros)
+		return;
+
+	if (icingadbName.IsEmpty()) {
+		ReportIcingadbCheck(checkable, commandObj, cr, "Icinga DB UNKNOWN: Attribute 'icingadb_name' must be set.", ServiceUnknown);
+		return;
+	}
+
+	auto conn (IcingaDB::GetByName(icingadbName));
+
+	if (!conn) {
+		ReportIcingadbCheck(checkable, commandObj, cr, "Icinga DB UNKNOWN: Icinga DB connection '" + icingadbName + "' does not exist.", ServiceUnknown);
+		return;
+	}
+
+	auto redis (conn->GetConnection());
+
+	if (!redis->GetConnected()) {
+		ReportIcingadbCheck(checkable, commandObj, cr, "Icinga DB CRITICAL: Could not connect to Redis.", ServiceCritical);
+		return;
+	}
+
+	auto now (Utility::GetTime());
+	Array::Ptr redisTime, xReadHeartbeat, xReadStats, xReadRtuHistory;
+
+	try {
+		auto replies (redis->GetResultsOfQueries(
+			{
+				{"TIME"},
+				{"XREAD", "STREAMS", "icingadb:telemetry:heartbeat", "0-0"},
+				{"XREAD", "STREAMS", "icingadb:telemetry:stats", "0-0"},
+				{
+					"XREAD", "COUNT", "1", "STREAMS",
+					"icinga:runtime", "icinga:runtime:state",
+					"icinga:history:stream:acknowledgement", "icinga:history:stream:comment",
+					"icinga:history:stream:downtime", "icinga:history:stream:flapping",
+					"icinga:history:stream:notification", "icinga:history:stream:state",
+					"0-0", "0-0", "0-0", "0-0", "0-0", "0-0", "0-0", "0-0"
+				}
```

Comment:
Looks like not all of these streams (in particular `icinga:runtime:state`) are drained by a standby Icinga DB instance, so the check further down gives me false warning on my second instance: `Query backlog: 1354.154 seconds, greater than CRITICAL threshold (900)!`